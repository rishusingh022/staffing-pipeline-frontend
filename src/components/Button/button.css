.button {
  background-color: #2251ff;
  border: none;
  color:#ffffff;
  padding: 11.5px 12px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 14px;
}
