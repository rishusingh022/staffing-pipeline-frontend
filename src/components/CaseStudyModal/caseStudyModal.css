.modal-content {
  background-color: white;
  width: 60%;
  height: 95%;
  overflow: auto;
}

.name {
  height: 13%;
}

.modal-image {
  height: 13%;
  display: flex;
}

.modal-input {
  padding-left: 30px;
}

.description {
  height: 13%;
}
.box-link {
  height: 13%;
}

.modal-collaborators {
  height: 80%;
  display: flex;
  flex-wrap: wrap;
  align-content: flex-start;
  overflow: scroll;
  gap: 20px;
}
.modal-engagement-name {
  height: 80%;
  display: flex;
  flex-wrap: wrap;
  align-content: flex-start;
  overflow: scroll;
  gap: 20px;
}

.enter-btn {
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  background-color: white;
}

.btn-comp {
  display: flex;
  flex-direction: row;
}

.no-users-involved {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal-search .search-container input {
  --tw-border-opacity: 1;
  border-color: rgb(156 163 175 / var(--tw-border-opacity));
}
.modal-user-search .search-container input {
  --tw-border-opacity: 1;
  border-color: rgb(156 163 175 / var(--tw-border-opacity));
}
.modal-title {
  font-weight: bolder;
}
