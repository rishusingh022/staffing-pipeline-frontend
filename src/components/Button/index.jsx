import React from 'react';
import './button.css';
import PropTypes from 'prop-types';

export default function Button({ buttonText, handleClick }) {
  return (
    <button onClick={handleClick} className="button" data-testid="button">
      {buttonText}
    </button>
  );
}

Button.propTypes = {
  buttonText: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired,
};
