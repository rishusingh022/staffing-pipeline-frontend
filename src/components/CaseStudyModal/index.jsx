import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import './caseStudyModal.css';
import PeopleHorizontalCard from '../PeopleHorizontalCard';
import Button from '../Button';
import EngagementHorizontalCard from '../EngagementHorizontalCard';
import makeRequest from '../../utils/makeRequest';
import Search from '../Search';
import {
  CREATE_CASE_STUDIES_DATA_URL,
  GET_ENGAGEMENT_DATA_URL,
  GET_USERS_INVOLVED_IN_ENGAGEMENT,
  UPLOAD_CASE_STUDY_IMAGE_ROUTE,
} from '../../constants/apiEndpoints';
import { useNavigate, useParams } from 'react-router-dom';

export default function CaseStudyModal({ setIsOpen }) {
  const { projectId } = useParams();
  const [allEngagements, setAllEngagements] = useState([]);
  const [selectedEngagement, setSelectedEngagement] = useState(projectId);
  const [users, setUsers] = useState([]);
  const [selectedUsers, setSelectedUsers] = useState([]);
  const [file, setFile] = useState(null);
  const [searchTerm, setSearchTerm] = useState('');
  const [userSearch, setUserSearch] = useState('');
  const [inputData, setInputdata] = useState({
    caseStudyname: '',
    description: '',
    boxLink: '',
  });
  const navigate = useNavigate();

  const filteredEngagements = allEngagements.filter(engagement => {
    if (searchTerm === '') {
      return true;
    }
    return engagement.name.toLowerCase().includes(searchTerm.toLowerCase());
  });

  const filteredUsers = users.filter(user => {
    if (userSearch === '') {
      return true;
    }
    return user.name.toLowerCase().includes(userSearch.toLowerCase());
  });

  const handleSearch = value => {
    setSearchTerm(value);
  };

  const handleUserSearch = value => {
    setUserSearch(value);
  };

  const handleSelectedEngagementChange = engagementId => {
    setSelectedUsers([]);
    setSelectedEngagement(engagementId);
  };

  useEffect(() => {
    makeRequest(GET_ENGAGEMENT_DATA_URL, {}, navigate)
      .then(response => {
        setAllEngagements(response);
      })
      .catch(err => {
        console.log(err);
      });
  }, []);

  useEffect(() => {
    makeRequest(GET_USERS_INVOLVED_IN_ENGAGEMENT(selectedEngagement), {}, navigate)
      .then(response => {
        setUsers(response);
      })
      .catch(err => {
        console.log(err);
      });
  }, [selectedEngagement]);

  const handleCaseStudyImage = async () => {
    const formData = new FormData();
    formData.append('file', file);
    return await makeRequest(
      UPLOAD_CASE_STUDY_IMAGE_ROUTE,
      {
        data: formData,
      },
      navigate
    );
  };

  const uploadCaseStudy = async newCaseStudy => {
    makeRequest(
      CREATE_CASE_STUDIES_DATA_URL,
      {
        data: newCaseStudy,
      },
      navigate
    )
      .then(() => {
        window.location.reload(false);
      })
      .catch(err => {
        console.log(err);
      });
  };
  const handleCreateCaseStudy = async () => {
    if (
      file &&
      inputData.caseStudyname &&
      inputData.description &&
      inputData.boxLink &&
      selectedEngagement &&
      selectedUsers.length > 0
    ) {
      const image = await handleCaseStudyImage();
      const newCaseStudy = {
        name: inputData.caseStudyname,
        description: inputData.description,
        boxLink: inputData.boxLink,
        engagementId: selectedEngagement,
        collaboratorsIds: selectedUsers,
        image: image.imageUrl,
      };
      await uploadCaseStudy(newCaseStudy);
      setIsOpen(false);
    } else {
      alert('Fields are empty.');
    }
  };

  return (
    <>
      <div className="background fixed bg-black/[.85] w-screen h-screen top-0 left-0 flex items-center justify-center">
        <div className="modal-content flex flex-col gap-4 p-9">
          <p className="modal-title">Upload Case Study</p>
          <div className="name">
            <input
              className="w-full px-3 py-2 border-[1px] border-gray-400"
              placeholder="Enter case study name"
              value={inputData.caseStudyname}
              onChange={event =>
                setInputdata({
                  ...inputData,
                  caseStudyname: event.target.value,
                })
              }
            />
          </div>
          <div className="modal-search">
            <Search placeHolderValue={'Search Engagement'} handleSearch={handleSearch} />
          </div>
          <div className="modal-engagement-name">
            {filteredEngagements.length > 0 &&
              filteredEngagements.map((data, index) => (
                <div className="w-[48.5%] h-[90px]" key={index}>
                  <EngagementHorizontalCard
                    engagementTitle={data.name}
                    engagementId={data.engagementId}
                    engagementImage={data.image}
                    selectedEngagement={selectedEngagement}
                    setSelectedEngagement={handleSelectedEngagementChange}
                  />
                </div>
              ))}
          </div>
          <div className="description">
            <input
              className="w-full px-3 py-2 border-[1px] border-gray-400"
              placeholder="Enter description"
              value={inputData.description}
              onChange={event =>
                setInputdata({
                  ...inputData,
                  description: event.target.value,
                })
              }
            />
          </div>
          <div className="box-link">
            <input
              className="w-full px-3 py-2 border-[1px] border-gray-400"
              placeholder="Paste Box Link"
              value={inputData.boxLink}
              onChange={event =>
                setInputdata({
                  ...inputData,
                  boxLink: event.target.value,
                })
              }
            />
          </div>
          <div className="modal-user-search">
            <Search placeHolderValue={'Search People'} handleSearch={handleUserSearch} />
          </div>
          <div className="modal-collaborators">
            {filteredUsers.length > 0 ? (
              filteredUsers.map((data, index) => (
                <div className="w-[48.5%] h-[90px] gap-2" key={index}>
                  <PeopleHorizontalCard
                    userFMNO={data.fmno}
                    userId={data.userId}
                    userName={data.name}
                    userImage={data.image}
                    userPosition={data.role}
                    userOffice="Bangalore"
                    selectedUsers={selectedUsers}
                    setSelectedUsers={setSelectedUsers}
                    selected={selectedUsers.includes(data.userId)}
                  />
                </div>
              ))
            ) : (
              <p className="no-users-involved font-light text-gray text-center">
                No users involved in this engagements.
              </p>
            )}
          </div>

          <div className="enter-btn p-0">
            <div className="btn-comp">
              <div className="submit-btn">
                <Button buttonText={'Cancel'} handleClick={() => setIsOpen(false)} />
              </div>
              <div className="modal-image">
                <label htmlFor="upload-case-study-image">
                  <p className="px-3 py-2 ml-2 bg-[#051b2c] text-white w-fit hover:cursor-pointer">Select Image</p>
                </label>
                <input
                  type={'file'}
                  name="Upload image"
                  id="upload-case-study-image"
                  className="hidden"
                  onChange={e => setFile(e.target.files[0])}
                />
                {file && <p className="mt-2 ml-2">Selected File: {file.name}</p>}
              </div>
            </div>
            <div className="cancel-btn">
              <Button buttonText={'Submit'} handleClick={handleCreateCaseStudy} />
            </div>
          </div>
        </div>
      </div>
    </>
  );
}

CaseStudyModal.propTypes = {
  setIsOpen: PropTypes.bool.isRequired,
};
