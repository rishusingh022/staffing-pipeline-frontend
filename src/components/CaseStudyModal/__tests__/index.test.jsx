import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import CaseStudyModal from '..';
import Button from '../../Button';

const mockedNavigate = jest.fn();
const mockedParam = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockedNavigate,
  useParams: () => mockedParam,
}));

describe('Case study modal', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<CaseStudyModal setIsOpen={true} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should call handleCreateCaseStudy function when clicked on submit button', () => {
    const handleCreateCaseStudy = jest.fn();
    render(<Button buttonText="Submit" handleClick={handleCreateCaseStudy} />);
    const button = screen.getByTestId('button');
    fireEvent.click(button);
    expect(handleCreateCaseStudy).toBeCalled();
  });
  it('should set setIsOpen to false when clicked on cancel button', () => {
    const setIsOpen = jest.fn();
    render(<Button buttonText="Cancel" handleClick={setIsOpen} />);
    const button = screen.getByTestId('button');
    fireEvent.click(button);
    expect(setIsOpen).toBeCalled();
  });
});
