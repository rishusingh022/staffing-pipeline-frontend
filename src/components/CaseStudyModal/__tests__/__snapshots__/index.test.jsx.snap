// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case study modal should render correctly 1`] = `
<DocumentFragment>
  <div
    class="background fixed bg-black/[.85] w-screen h-screen top-0 left-0 flex items-center justify-center z-10"
  >
    <div
      class="modal-content flex flex-col gap-4 p-9"
    >
      <p
        class="modal-title"
      >
        Upload Case Study
      </p>
      <div
        class="name"
      >
        <input
          class="w-full px-3 py-2 border-[1px] border-gray-400"
          placeholder="Enter case study name"
          value=""
        />
      </div>
      <div
        class="modal-search"
      >
        <div
          class="search-container"
        >
          <input
            placeholder="Search Engagement"
            type="text"
          />
          <div
            class="filter-search-button"
          >
            <button
              class="button"
              data-testid="button"
            >
              Search
            </button>
          </div>
        </div>
      </div>
      <div
        class="modal-engagement-name"
      />
      <div
        class="description"
      >
        <input
          class="w-full px-3 py-2 border-[1px] border-gray-400"
          placeholder="Enter description"
          value=""
        />
      </div>
      <div
        class="box-link"
      >
        <input
          class="w-full px-3 py-2 border-[1px] border-gray-400"
          placeholder="Paste Box Link"
          value=""
        />
      </div>
      <div
        class="modal-user-search"
      >
        <div
          class="search-container"
        >
          <input
            placeholder="Search People"
            type="text"
          />
          <div
            class="filter-search-button"
          >
            <button
              class="button"
              data-testid="button"
            >
              Search
            </button>
          </div>
        </div>
      </div>
      <div
        class="modal-collaborators"
      >
        <p
          class="no-users-involved font-light text-gray text-center"
        >
          No users involved in this Engagement
        </p>
      </div>
      <div
        class="enter-btn p-0"
      >
        <div
          class="btn-comp"
        >
          <div
            class="submit-btn"
          >
            <button
              class="button"
              data-testid="button"
            >
              Cancel
            </button>
          </div>
          <div
            class="modal-image"
          >
            <label
              for="upload-case-study-image"
            >
              <p
                class="px-3 py-2 ml-2 bg-[#051b2c] text-white w-fit hover:cursor-pointer"
              >
                Select Image
              </p>
            </label>
            <input
              class="hidden"
              id="upload-case-study-image"
              name="Upload image"
              type="file"
            />
          </div>
        </div>
        <div
          class="cancel-btn"
        >
          <button
            class="button"
            data-testid="button"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
